push-payload/index.js

```diff
@@ -2,20 +2,51 @@ var endpoint;
 var key;
 var authSecret;
 
+function urlBase64ToUint8Array(base64String) {
+  const padding = '='.repeat((4 - base64String.length % 4) % 4);
+  const base64 = (base64String + padding)
+    .replace(/\-/g, '+')
+    .replace(/_/g, '/');
+ 
+  const rawData = window.atob(base64);
+  const outputArray = new Uint8Array(rawData.length);
+ 
+  for (let i = 0; i < rawData.length; ++i) {
+    outputArray[i] = rawData.charCodeAt(i);
+  }
+  return outputArray;
+}
+
 // Register a Service Worker.
 navigator.serviceWorker.register('service-worker.js')
 .then(function(registration) {
   // Use the PushManager to get the user's subscription to the push service.
   return registration.pushManager.getSubscription()
-  .then(function(subscription) {
-    // If a subscription was found, return it.
+  .then(async function(subscription) {
+    // Get the server's public key
+    var response = await fetch('./vapidPublicKey')
+    var vapidPublicKey = await response.text();
+    // Convert the URL safe base64 string to a Uint8Array
+    var convertedVapidKey = urlBase64ToUint8Array(vapidPublicKey);
+
+    // See if we already have a subscription with the server
     if (subscription) {
-      return subscription;
+      var currentKey = new Uint8Array(subscription.options.applicationServerKey);
```

Comment:
I'm not sure that we need to take into account the possibility of the vapid key changing in these examples.

Reply:
If we don't store the key somewhere (as the example is right now), it will change every time the server process restarts. When the key has changed the example will no longer work: `server.js` will attempt to send the push message and the endpoint will return an 400 UnauthorizedRegistration response (Google) or 401 Unauthorized response with "Request did not validate missing authorization header" (Mozilla). 

This obviously happens a lot when developing locally. I'm not sure what kind of infrastructure serviceworke.rs is running on, but say you try the example one day, the server restarts overnight, and you try it again the following day, it will no longer work.

Reply:
We are using Heroku. I think we can store the VAPID key in a file so it doesn't get recreated every time.
It will still be recreated from time to time, but not every time you reload.

Reply:
It is still going to be a pain though... Hopefully we can figure out a better solution.

Reply:
I still think the current check is a pretty good solution. Even in production scenarios you might need to change the keys sometimes (when the key is compromised or lost). This seems a sensible place to check if the subscription we have is still usable. And as a bonus we don't have to extend the demo with persistent storage (or something behind-the-scenes specifically for serviceworke.rs).

Reply:
I have another idea. We can put the VAPID key in a Heroku environment variable, so it will never change.